File: leetcode_py/Sort List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def merge(self, p, q):
        head = ListNode(0)
        pre = head
        while p and q:
            if p.val <= q.val:
                pre.next = p
                p = p.next
            else:
                pre.next = q
                q = q.next
            pre = pre.next
        if p:
            pre.next = p
        elif q:
            pre.next = q
        return head.next

    # @param {ListNode} head
    # @return {ListNode}
    def sortList(self, head):
        if not head or not head.next:
            return head
        p, q, p_end = head, head, None
        while q and q.next:
            p_end = p
            p = p.next
            q = q.next.next
        p_end.next = None
        q = self.sortList(p)
        p = self.sortList(head)

        return self.merge(p, q)


class Solution:
    # @param head, a ListNode
    # @return a ListNode
    def sortList(self, head):
        if head == None or head.next == None:
            return head

        left = head
        right = head
        while right.next:
            right = right.next
            if right.next:
                right = right.next
                left = left.next
            else:
                break
        tmp = left.next
        left.next = None
        left = self.sortList(head)
        right = self.sortList(tmp)
        return self.merge(left, right)

    def merge(self, left, right):
        if right == None or left == None:
            return left
        dummy0 = ListNode(0)
        pre = dummy0
        pre.next = left
        dummy = ListNode(0)
        pre2 = dummy
        dummy.next = right

        while left and right:
            if left.val <= right.val:
                pre = left
                left = left.next
            else:
                pre2.next = right.next
                right.next = pre.next
                pre.next = right
                pre = right
                right = pre2.next
        if right:
            pre.next = right
        return dummy0.next

# https://leetcode.com/discuss/15420/java-solution-with-strict-o-1-auxiliary-space-complexity
# https://leetcode.com/discuss/28594/bottom-recurring-with-space-complextity-time-complextity

File: leetcode_js/Sort List.js
//148. Sort List My Submissions Question
//Total Accepted: 61309 Total Submissions: 258737 Difficulty: Medium
//Sort a linked list in O(n log n) time using constant space complexity.
//
function ListNode(val) {
      this.val = val;
      this.next = null;
 }
function merge(p, q){
    var head = new ListNode(0), pidx = 0 , qidx = 0, tmphead = head;
    while(p && q){
        if(p.val <= q.val){
            head.next = p;
            p = p.next;
        }else{
            head.next = q;
            q = q.next;
        }
        head = head.next;
    }
    while(p){
        head.next = p;
        p = p.next;
        head = head.next;
    }
    while(q){
        head.next = q;
        q = q.next;
        head = head.next
    }
    return tmphead.next;
}
var sortList = function(head) {
    if(!head || !head.next) return head;
    var p1 = head, p2 = p1.next;
    while(p2.next && p2.next.next){
        p1 = p1.next;
        p2 = p2.next.next;
    } 
    p2 = sortList(p1.next);
    p1.next = null;
    p1 = sortList(head);  
    return merge(p1,p2);
};
hh = new ListNode(3);
hh.next = new ListNode(2);
hh.next.next = new ListNode(4);
var tmp = sortList(hh);
while(tmp){
document.write(tmp.val);
tmp = tmp.next;
}